### ISSUE TITLE
[placement group] Resource leak when allocating failure

### ISSUE THREAD

**Issue by iycheng** [ray-project/ray#17044]
### What is the problem?

*Ray version and other system information (Python version, TensorFlow version, OS):*
When allocating resource failed, sometimes the resource failed to be released, ([code](https://github.com/ray-project/ray/blob/1fd0eb805e49c937f4670cb436fdddec6a3c158f/src/ray/raylet/scheduling/cluster_resource_scheduler.cc#L807-L809)) which might leads to a leak.

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

**Comment by rkooo567** [comment]
I have a couple questions; 

- Without this fix, is your PR going to be broken? How did you find the issue?
- Also, can you update the description? (detailed problem)

**Comment by iycheng** [comment]
@rkooo567  No, my PR is not going to be broken. The bug is identified by eyeball.
Btw, if it's about leak, I'll say it's an important issue, should we add it back to P1?

**Comment by wuisawesome** [comment]
I think we might've partially allocated the resources, so we leave it up to the caller of that function to release the resources since `task_allocation` is still properly set.

**Comment by rkooo567** [comment]
@wuisawesome this means if the caller forgets to release resources, it can have a leak right? This seems to be pretty dangerous API?

**Comment by wuisawesome** [comment]
yeah
